File: sicora-be-python/evalinservice/app/infrastructure/adapters/user_service_adapter.py
"""User Service adapter."""

from typing import Optional
from uuid import UUID
import httpx

from ...application.interfaces import UserServiceInterface


class UserServiceAdapter(UserServiceInterface):
    """
    Adaptador para comunicación con UserService.
    
    Responsabilidades:
    - Implementar interfaz de UserService 
    - Realizar llamadas HTTP al microservicio UserService
    - Manejar errores de comunicación
    """
    
    def __init__(self, base_url: str):
        self._base_url = base_url.rstrip("/")
        self._client = httpx.AsyncClient()
    
    async def get_user_by_id(self, user_id: UUID) -> Optional[dict]:
        """
        Obtener información de usuario por ID.
        
        Args:
            user_id: ID del usuario
            
        Returns:
            dict: Información del usuario o None si no existe
        """
        try:
            response = await self._client.get(f"{self._base_url}/api/v1/users/{user_id}")
            
            if response.status_code == 200:
                return response.json()
            elif response.status_code == 404:
                return None
            else:
                response.raise_for_status()
                
        except httpx.RequestError as e:
            # Log error en producción
            print(f"Error al conectar con UserService: {e}")
            return None
        
        return None
    
    async def get_users_by_role(self, role: str) -> list:
        """
        Obtener usuarios por rol.
        
        Args:
            role: Rol de los usuarios a buscar
            
        Returns:
            list: Lista de usuarios con el rol especificado
        """
        try:
            response = await self._client.get(
                f"{self._base_url}/api/v1/users",
                params={"role": role}
            )
            
            if response.status_code == 200:
                return response.json()
            else:
                response.raise_for_status()
                
        except httpx.RequestError as e:
            print(f"Error al conectar con UserService: {e}")
            return []
        
        return []
    
    async def user_exists(self, user_id: UUID) -> bool:
        """
        Verificar si un usuario existe.
        
        Args:
            user_id: ID del usuario
            
        Returns:
            bool: True si el usuario existe
        """
        user = await self.get_user_by_id(user_id)
        return user is not None
    
    async def close(self):
        """Cerrar el cliente HTTP."""
        await self._client.aclose()

File: sicora-be-go/evalinservice/internal/infrastructure/external/user_service_adapter.go
package external

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"

	"github.com/google/uuid"
)

// UserServiceAdapter representa el adaptador para el servicio de usuarios
type UserServiceAdapter struct {
	baseURL    string
	httpClient *http.Client
	apiKey     string
}

// UserInfo representa la información básica de un usuario
type UserInfo struct {
	ID        uuid.UUID `json:"id"`
	Email     string    `json:"email"`
	FirstName string    `json:"first_name"`
	LastName  string    `json:"last_name"`
	Role      string    `json:"role"`
	IsActive  bool      `json:"is_active"`
}

// UserResponse representa la respuesta del servicio de usuarios
type UserResponse struct {
	Data    *UserInfo `json:"data"`
	Message string    `json:"message"`
	Success bool      `json:"success"`
}

// UsersResponse representa la respuesta para múltiples usuarios
type UsersResponse struct {
	Data    []*UserInfo `json:"data"`
	Message string      `json:"message"`
	Success bool        `json:"success"`
}

// NewUserServiceAdapter crea un nuevo adaptador para el servicio de usuarios
func NewUserServiceAdapter(baseURL, apiKey string) *UserServiceAdapter {
	return &UserServiceAdapter{
		baseURL: baseURL,
		apiKey:  apiKey,
		httpClient: &http.Client{
			Timeout: 30 * time.Second,
		},
	}
}

// GetUserByID obtiene un usuario por su ID
func (a *UserServiceAdapter) GetUserByID(ctx context.Context, userID uuid.UUID) (*UserInfo, error) {
	url := fmt.Sprintf("%s/api/v1/users/%s", a.baseURL, userID.String())

	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Authorization", "Bearer "+a.apiKey)
	req.Header.Set("Content-Type", "application/json")

	resp, err := a.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to make request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("user service returned status %d", resp.StatusCode)
	}

	var response UserResponse
	if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	if !response.Success {
		return nil, fmt.Errorf("user service error: %s", response.Message)
	}

	return response.Data, nil
}

// GetUsersByRole obtiene usuarios por rol
func (a *UserServiceAdapter) GetUsersByRole(ctx context.Context, role string) ([]*UserInfo, error) {
	url := fmt.Sprintf("%s/api/v1/users?role=%s", a.baseURL, role)

	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Authorization", "Bearer "+a.apiKey)
	req.Header.Set("Content-Type", "application/json")

	resp, err := a.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to make request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("user service returned status %d", resp.StatusCode)
	}

	var response UsersResponse
	if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	if !response.Success {
		return nil, fmt.Errorf("user service error: %s", response.Message)
	}

	return response.Data, nil
}

// ValidateUser valida si un usuario existe y está activo
func (a *UserServiceAdapter) ValidateUser(ctx context.Context, userID uuid.UUID) (bool, error) {
	user, err := a.GetUserByID(ctx, userID)
	if err != nil {
		return false, err
	}

	return user != nil && user.IsActive, nil
}

// GetInstructors obtiene todos los instructores
func (a *UserServiceAdapter) GetInstructors(ctx context.Context) ([]*UserInfo, error) {
	return a.GetUsersByRole(ctx, "instructor")
}

// GetStudents obtiene todos los estudiantes
func (a *UserServiceAdapter) GetStudents(ctx context.Context) ([]*UserInfo, error) {
	return a.GetUsersByRole(ctx, "student")
}

// GetAdministrators obtiene todos los administradores
func (a *UserServiceAdapter) GetAdministrators(ctx context.Context) ([]*UserInfo, error) {
	return a.GetUsersByRole(ctx, "admin")
}
